File: clients/ios/Lextures/Core/LMS/CourseSettingsLogic.swift
import Foundation

/// Course settings helpers (M13.1) — permission gate, validation, dirty detection, schedule/theme helpers.
enum CourseSettingsLogic {
    static func courseItemCreatePermission(courseCode: String) -> String {
        "course:\(courseCode):item:create"
    }

    static func canManageCourse(courseCode: String, permissions: [String]) -> Bool {
        permissions.contains(courseItemCreatePermission(courseCode: courseCode))
    }

    static func settingsEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffMobileCourseSettings
    }

    static func shouldShowWorkspaceSection(
        course: CourseSummary,
        permissions: [String],
        features: MobilePlatformFeatures
    ) -> Bool {
        guard settingsEnabled(features) else { return false }
        return canManageCourse(courseCode: course.courseCode, permissions: permissions)
    }

    enum CourseSettingsSection: String, CaseIterable, Identifiable, Hashable {
        case general
        case features
        case sections
        case grading
        case outcomes
        case gradingAgents
        case plagiarism
        case accessibility
        case translations
        case importExport
        case blueprint
        case archive

        var id: String { rawValue }

        var labelKey: String {
            switch self {
            case .general: return "mobile.courseSettings.section.general"
            case .features: return "mobile.courseSettings.section.features"
            case .sections: return "mobile.courseSettings.section.sections"
            case .grading: return "mobile.courseSettings.section.grading"
            case .outcomes: return "mobile.courseSettings.section.outcomes"
            case .gradingAgents: return "mobile.courseSettings.section.gradingAgents"
            case .plagiarism: return "mobile.courseSettings.section.plagiarism"
            case .accessibility: return "mobile.courseSettings.section.accessibility"
            case .translations: return "mobile.courseSettings.section.translations"
            case .importExport: return "mobile.courseSettings.section.importExport"
            case .blueprint: return "mobile.courseSettings.section.blueprint"
            case .archive: return "mobile.courseSettings.section.archive"
            }
        }

        var label: String { L.text(String.LocalizationValue(labelKey)) }

        var systemImage: String {
            switch self {
            case .general: return "info.circle"
            case .features: return "slider.horizontal.3"
            case .sections: return "square.grid.2x2"
            case .grading: return "scalemass"
            case .outcomes: return "target"
            case .gradingAgents: return "cpu"
            case .plagiarism: return "shield"
            case .accessibility: return "eye"
            case .translations: return "globe"
            case .importExport: return "arrow.up.arrow.down"
            case .blueprint: return "doc.on.doc"
            case .archive: return "archivebox"
            }
        }
    }

    static func visibleSettingsSections(
        course: CourseSummary,
        features: MobilePlatformFeatures
    ) -> [CourseSettingsSection] {
        var sections: [CourseSettingsSection] = [.general, .features]
        if course.isSectionsEnabled {
            sections.append(.sections)
        }
        sections.append(contentsOf: [.grading, .outcomes])
        if features.graderAgentEnabled {
            sections.append(.gradingAgents)
        }
        if features.ffPlagiarismChecks {
            sections.append(.plagiarism)
        }
        if features.altTextEnforcementEnabled || features.translationMemoryEnabled {
            if features.altTextEnforcementEnabled {
                sections.append(.accessibility)
            }
            if features.translationMemoryEnabled {
                sections.append(.translations)
            }
        }
        sections.append(contentsOf: [.importExport, .blueprint, .archive])
        return sections
    }

    static let gradeLevels: [String] = [
        "", "K", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12",
        "K-2", "3-5", "6-8", "9-12", "K-12",
    ]

    static let markdownThemePresets: [String] = [
        "classic", "reader", "serif", "contrast", "night", "accent", "custom",
    ]

    static let articleWidths: [String] = ["narrow", "comfortable", "wide", "full"]
    static let fontFamilies: [String] = ["sans", "serif"]

    enum RelativeDurationUnit: String, CaseIterable {
        case days = "D"
        case weeks = "W"
        case months = "M"
        case years = "Y"
    }

    enum CourseHomeLanding: String, CaseIterable {
        case data, calendar
        case contentPage = "content_page"
    }

    enum ScheduleMode: String, CaseIterable {
        case fixed, relative
    }

    enum SaveStatus: Equatable {
        case idle, saving, saved, error(String)
    }

    struct ValidationError: Equatable {
        var title: String?
        var courseHome: String?
    }

    static func validateGeneralForm(
        title: String,
        courseHomeLanding: CourseHomeLanding,
        courseHomeContentItemId: String
    ) -> ValidationError? {
        var error = ValidationError()
        if title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            error.title = L.text("mobile.courseSettings.validation.titleRequired")
        }
        if courseHomeLanding == .contentPage,
           courseHomeContentItemId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            error.courseHome = L.text("mobile.courseSettings.validation.contentPageRequired")
        }
        if error.title == nil && error.courseHome == nil { return nil }
        return error
    }

    static func normalizeCourseHomeLanding(_ value: String?) -> CourseHomeLanding {
        if value == CourseHomeLanding.calendar.rawValue { return .calendar }
        if value == CourseHomeLanding.contentPage.rawValue { return .contentPage }
        return .data
    }

    static func isoDurationToParts(
        iso: String?
    ) -> (amount: String, unit: RelativeDurationUnit) {
        guard let iso, !iso.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return ("", .months)
        }
        let pattern = /^P(\d+)([DWMY])$/
        guard let match = iso.trimmingCharacters(in: .whitespacesAndNewlines).wholeMatch(of: pattern) else {
            return ("", .months)
        }
        let amount = String(match.output.1)
        let unitChar = String(match.output.2).uppercased()
        let unit = RelativeDurationUnit(rawValue: unitChar) ?? .months
        return (amount, unit)
    }

    static func partsToIsoDuration(amount: String, unit: RelativeDurationUnit) -> String? {
        guard let parsedAmount = Int(amount.trimmingCharacters(in: .whitespacesAndNewlines)), parsedAmount >= 1 else {
            return nil
        }
        return "P\(parsedAmount)\(unit.rawValue)"
    }

    static func isoToLocalDateString(_ iso: String?) -> String {
        guard let iso, !iso.isEmpty, let date = ISO8601DateFormatter().date(from: iso) else {
            return ""
        }
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm"
        formatter.timeZone = .current
        return formatter.string(from: date)
    }

    static func localDateStringToIso(_ value: String) -> String? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm"
        formatter.timeZone = .current
        guard let date = formatter.date(from: trimmed) else { return nil }
        return ISO8601DateFormatter().string(from: date)
    }

    static func normalizeIso(_ iso: String?) -> String? {
        localDateStringToIso(isoToLocalDateString(iso))
    }

    static func parseHeroObjectPosition(_ pos: String?) -> (x: Double, y: Double) {
        guard let pos, !pos.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return (50, 50)
        }
        let pattern = /^(\d+(?:\.\d+)?)%\s+(\d+(?:\.\d+)?)%$/
        guard let match = pos.trimmingCharacters(in: .whitespacesAndNewlines).wholeMatch(of: pattern) else {
            return (50, 50)
        }
        let posX = min(100, max(0, Double(match.output.1) ?? 50))
        let posY = min(100, max(0, Double(match.output.2) ?? 50))
        return (posX, posY)
    }

    static func formatHeroObjectPosition(x: Double, y: Double) -> String? {
        let rx = round(x)
        let ry = round(y)
        if rx == 50 && ry == 50 { return nil }
        return "\(Int(rx))% \(Int(ry))%"
    }

    static func defaultImagePrompt(title: String, description: String) -> String {
        """
        Generate an image for a course banner with the following title and description:
        Title: \(title)
        Description: \(description)
        """
    }

    static func gradeLevelLabel(_ value: String) -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return L.text("mobile.courseSettings.gradeLevel.none") }
        return trimmed
    }

    static func timezoneOptions() -> [String] {
        TimeZone.knownTimeZoneIdentifiers.sorted()
    }

    static func defaultTimezone() -> String {
        TimeZone.current.identifier
    }

    static func isGeneralFormDirty(form: CourseGeneralFormState, course: CourseSummary) -> Bool {
        isBasicInfoDirty(form: form, course: course)
            || isCourseHomeDirty(form: form, course: course)
            || isScheduleDirty(form: form, course: course)
            || isMarkdownThemeDirty(form: form, course: course)
    }

    private static func isBasicInfoDirty(form: CourseGeneralFormState, course: CourseSummary) -> Bool {
        if form.title.trimmingCharacters(in: .whitespacesAndNewlines) != course.title { return true }
        if form.description.trimmingCharacters(in: .whitespacesAndNewlines) != course.description { return true }
        if form.published != (course.published ?? false) { return true }
        if (form.gradeLevel.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty)
            != (course.gradeLevel?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty) {
            return true
        }
        if (form.courseTimezone.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty)
            != (course.courseTimezone?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
                ?? defaultTimezone().nilIfEmpty) {
            return true
        }
        return false
    }

    private static func isCourseHomeDirty(form: CourseGeneralFormState, course: CourseSummary) -> Bool {
        let normHome = normalizeCourseHomeLanding(course.courseHomeLanding)
        if form.courseHomeLanding != normHome { return true }
        let origHomeId = (course.courseHomeContentItemId ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        return form.courseHomeLanding == .contentPage
            && form.courseHomeContentItemId.trimmingCharacters(in: .whitespacesAndNewlines) != origHomeId
    }

    private static func isScheduleDirty(form: CourseGeneralFormState, course: CourseSummary) -> Bool {
        let mode = course.scheduleMode == ScheduleMode.relative.rawValue ? ScheduleMode.relative : .fixed
        if form.scheduleMode != mode { return true }
        if form.scheduleMode == .fixed {
            if localDateStringToIso(form.startsAt) != normalizeIso(course.startsAt) { return true }
            if localDateStringToIso(form.endsAt) != normalizeIso(course.endsAt) { return true }
            if localDateStringToIso(form.visibleFrom) != normalizeIso(course.visibleFrom) { return true }
            if localDateStringToIso(form.hiddenAt) != normalizeIso(course.hiddenAt) { return true }
            return false
        }
        let currentEnd = partsToIsoDuration(amount: form.relEndAmount, unit: form.relEndUnit)
        if currentEnd != (course.relativeEndAfter?.nilIfEmpty) { return true }
        let currentHidden = partsToIsoDuration(amount: form.relHiddenAmount, unit: form.relHiddenUnit)
        return currentHidden != (course.relativeHiddenAfter?.nilIfEmpty)
    }

    private static func isMarkdownThemeDirty(form: CourseGeneralFormState, course: CourseSummary) -> Bool {
        if (course.markdownThemePreset ?? "default") != form.markdownThemePreset { return true }
        guard form.markdownThemePreset == "custom" else { return false }
        let seed = MarkdownThemeCustom.seed
        let orig = course.markdownThemeCustom ?? seed
        let draft = form.customDraft
        if (draft.headingColor ?? seed.headingColor) != (orig.headingColor ?? seed.headingColor) { return true }
        if (draft.bodyColor ?? seed.bodyColor) != (orig.bodyColor ?? seed.bodyColor) { return true }
        if (draft.linkColor ?? seed.linkColor) != (orig.linkColor ?? seed.linkColor) { return true }
        if (draft.codeBackground ?? seed.codeBackground) != (orig.codeBackground ?? seed.codeBackground) { return true }
        if (draft.blockquoteBorder ?? seed.blockquoteBorder) != (orig.blockquoteBorder ?? seed.blockquoteBorder) { return true }
        if (draft.articleWidth ?? seed.articleWidth) != (orig.articleWidth ?? seed.articleWidth) { return true }
        return (draft.fontFamily ?? seed.fontFamily) != (orig.fontFamily ?? seed.fontFamily)
    }

    static func applyCourseToForm(_ course: CourseSummary) -> CourseGeneralFormState {
        let endParts = isoDurationToParts(iso: course.relativeEndAfter)
        let hiddenParts = isoDurationToParts(iso: course.relativeHiddenAfter)
        return CourseGeneralFormState(
            title: course.title,
            description: course.description,
            published: course.published ?? false,
            gradeLevel: course.gradeLevel ?? "",
            courseTimezone: course.courseTimezone?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
                ?? defaultTimezone(),
            courseHomeLanding: normalizeCourseHomeLanding(course.courseHomeLanding),
            courseHomeContentItemId: (course.courseHomeContentItemId ?? "").trimmingCharacters(in: .whitespacesAndNewlines),
            scheduleMode: course.scheduleMode == ScheduleMode.relative.rawValue ? .relative : .fixed,
            startsAt: isoToLocalDateString(course.startsAt),
            endsAt: isoToLocalDateString(course.endsAt),
            visibleFrom: isoToLocalDateString(course.visibleFrom),
            hiddenAt: isoToLocalDateString(course.hiddenAt),
            relEndAmount: endParts.amount,
            relEndUnit: endParts.unit,
            relHiddenAmount: hiddenParts.amount,
            relHiddenUnit: hiddenParts.unit,
            markdownThemePreset: course.markdownThemePreset ?? "default",
            customDraft: {
                var draft = MarkdownThemeCustom.seed
                if let custom = course.markdownThemeCustom {
                    draft.merge(custom)
                }
                return draft
            }()
        )
    }

    static func buildCourseUpdateRequest(form: CourseGeneralFormState) -> CourseUpdateRequest {
        let mode = form.scheduleMode
        return CourseUpdateRequest(
            title: form.title.trimmingCharacters(in: .whitespacesAndNewlines),
            description: form.description.trimmingCharacters(in: .whitespacesAndNewlines),
            published: form.published,
            startsAt: mode == .relative ? nil : localDateStringToIso(form.startsAt),
            endsAt: mode == .relative ? nil : localDateStringToIso(form.endsAt),
            visibleFrom: mode == .relative ? nil : localDateStringToIso(form.visibleFrom),
            hiddenAt: mode == .relative ? nil : localDateStringToIso(form.hiddenAt),
            scheduleMode: mode.rawValue,
            relativeEndAfter: mode == .relative ? partsToIsoDuration(amount: form.relEndAmount, unit: form.relEndUnit) : nil,
            relativeHiddenAfter: mode == .relative ? partsToIsoDuration(amount: form.relHiddenAmount, unit: form.relHiddenUnit) : nil,
            courseHomeLanding: form.courseHomeLanding.rawValue,
            courseHomeContentItemId: form.courseHomeLanding == .contentPage
                ? form.courseHomeContentItemId.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
                : nil,
            courseTimezone: form.courseTimezone.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty,
            gradeLevel: form.gradeLevel.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
        )
    }

    static func buildMarkdownThemePatch(form: CourseGeneralFormState) -> CourseMarkdownThemePatch {
        CourseMarkdownThemePatch(
            preset: form.markdownThemePreset,
            custom: form.markdownThemePreset == "custom" ? form.customDraft : nil
        )
    }

    static func courseNeedsUpdate(form: CourseGeneralFormState, course: CourseSummary) -> Bool {
        let request = buildCourseUpdateRequest(form: form)
        let baseline = buildCourseUpdateRequest(form: applyCourseToForm(course))
        return request != baseline
    }

    static func themeNeedsUpdate(form: CourseGeneralFormState, course: CourseSummary) -> Bool {
        let request = buildMarkdownThemePatch(form: form)
        let baseline = buildMarkdownThemePatch(form: applyCourseToForm(course))
        return request != baseline
    }

    static func contentPages(from items: [CourseStructureItem]) -> [CourseStructureItem] {
        items.filter { $0.kind == "content_page" }
    }

    static func cacheKeySettings(courseCode: String) -> String {
        "course:\(courseCode):settings"
    }
}

private extension String {
    var nilIfEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

struct CourseGeneralFormState: Equatable {
    var title: String = ""
    var description: String = ""
    var published: Bool = false
    var gradeLevel: String = ""
    var courseTimezone: String = CourseSettingsLogic.defaultTimezone()
    var courseHomeLanding: CourseSettingsLogic.CourseHomeLanding = .data
    var courseHomeContentItemId: String = ""
    var scheduleMode: CourseSettingsLogic.ScheduleMode = .fixed
    var startsAt: String = ""
    var endsAt: String = ""
    var visibleFrom: String = ""
    var hiddenAt: String = ""
    var relEndAmount: String = ""
    var relEndUnit: CourseSettingsLogic.RelativeDurationUnit = .months
    var relHiddenAmount: String = ""
    var relHiddenUnit: CourseSettingsLogic.RelativeDurationUnit = .months
    var markdownThemePreset: String = "default"
    var customDraft: MarkdownThemeCustom = .seed
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/CourseSettingsLogic.kt
package com.lextures.android.core.lms

import java.util.TimeZone
import java.util.regex.Pattern

/** Course settings helpers (M13.1). */
object CourseSettingsLogic {
    fun courseItemCreatePermission(courseCode: String): String = "course:$courseCode:item:create"

    fun canManageCourse(courseCode: String, permissions: List<String>): Boolean =
        permissions.contains(courseItemCreatePermission(courseCode))

    fun settingsEnabled(features: com.lextures.android.core.navigation.MobilePlatformFeatures): Boolean =
        features.ffMobileCourseSettings

    fun shouldShowWorkspaceSection(
        course: CourseSummary,
        permissions: List<String>,
        features: com.lextures.android.core.navigation.MobilePlatformFeatures,
    ): Boolean {
        if (!settingsEnabled(features)) return false
        return canManageCourse(course.courseCode, permissions)
    }

    enum class CourseSettingsSection(val labelRes: String) {
        General("mobile_courseSettings_section_general"),
        Features("mobile_courseSettings_section_features"),
        Marketplace("mobile_courseSettings_section_marketplace"),
        Sections("mobile_courseSettings_section_sections"),
        Grading("mobile_courseSettings_section_grading"),
        Outcomes("mobile_courseSettings_section_outcomes"),
        GradingAgents("mobile_courseSettings_section_gradingAgents"),
        Plagiarism("mobile_courseSettings_section_plagiarism"),
        Accessibility("mobile_courseSettings_section_accessibility"),
        Translations("mobile_courseSettings_section_translations"),
        ImportExport("mobile_courseSettings_section_importExport"),
        Blueprint("mobile_courseSettings_section_blueprint"),
        Archive("mobile_courseSettings_section_archive"),
    }

    fun visibleSettingsSections(
        course: CourseSummary,
        features: com.lextures.android.core.navigation.MobilePlatformFeatures,
    ): List<CourseSettingsSection> = buildList {
        add(CourseSettingsSection.General)
        add(CourseSettingsSection.Features)
        if (features.ffCourseMarketplace) add(CourseSettingsSection.Marketplace)
        if (course.isSectionsEnabled) add(CourseSettingsSection.Sections)
        add(CourseSettingsSection.Grading)
        add(CourseSettingsSection.Outcomes)
        if (features.graderAgentEnabled) add(CourseSettingsSection.GradingAgents)
        if (features.ffPlagiarismChecks) add(CourseSettingsSection.Plagiarism)
        if (features.altTextEnforcementEnabled) add(CourseSettingsSection.Accessibility)
        if (features.translationMemoryEnabled) add(CourseSettingsSection.Translations)
        add(CourseSettingsSection.ImportExport)
        add(CourseSettingsSection.Blueprint)
        add(CourseSettingsSection.Archive)
    }

    val gradeLevels = listOf(
        "", "K", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12",
        "K-2", "3-5", "6-8", "9-12", "K-12",
    )

    val markdownThemePresets = listOf("classic", "reader", "serif", "contrast", "night", "accent", "custom")

    enum class RelativeDurationUnit { D, W, M, Y }

    enum class CourseHomeLanding { data, calendar, content_page }

    enum class ScheduleMode { fixed, relative }

    data class ValidationError(val title: String? = null, val courseHome: String? = null)

    fun validateGeneralForm(
        title: String,
        courseHomeLanding: CourseHomeLanding,
        courseHomeContentItemId: String,
    ): ValidationError? {
        var titleError: String? = null
        var homeError: String? = null
        if (title.trim().isEmpty()) titleError = "title-required"
        if (courseHomeLanding == CourseHomeLanding.content_page && courseHomeContentItemId.trim().isEmpty()) {
            homeError = "content-page-required"
        }
        if (titleError == null && homeError == null) return null
        return ValidationError(titleError, homeError)
    }

    fun normalizeCourseHomeLanding(value: String?): CourseHomeLanding = when (value) {
        "calendar" -> CourseHomeLanding.calendar
        "content_page" -> CourseHomeLanding.content_page
        else -> CourseHomeLanding.data
    }

    private val durationPattern = Pattern.compile("^P(\\d+)([DWMY])$", Pattern.CASE_INSENSITIVE)

    fun isoDurationToParts(iso: String?): Pair<String, RelativeDurationUnit> {
        if (iso.isNullOrBlank()) return "" to RelativeDurationUnit.M
        val matcher = durationPattern.matcher(iso.trim())
        if (!matcher.matches()) return "" to RelativeDurationUnit.M
        val unit = runCatching { RelativeDurationUnit.valueOf(matcher.group(2)!!.uppercase()) }
            .getOrDefault(RelativeDurationUnit.M)
        return matcher.group(1)!! to unit
    }

    fun partsToIsoDuration(amount: String, unit: RelativeDurationUnit): String? {
        val n = amount.trim().toIntOrNull() ?: return null
        if (n < 1) return null
        return "P$n$unit"
    }

    fun parseHeroObjectPosition(pos: String?): Pair<Double, Double> {
        if (pos.isNullOrBlank()) return 50.0 to 50.0
        val matcher = Pattern.compile("^(\\d+(?:\\.\\d+)?)%\\s+(\\d+(?:\\.\\d+)?)%$").matcher(pos.trim())
        if (!matcher.matches()) return 50.0 to 50.0
        val x = matcher.group(1)!!.toDoubleOrNull()?.coerceIn(0.0, 100.0) ?: 50.0
        val y = matcher.group(2)!!.toDoubleOrNull()?.coerceIn(0.0, 100.0) ?: 50.0
        return x to y
    }

    fun formatHeroObjectPosition(x: Double, y: Double): String? {
        val rx = kotlin.math.round(x)
        val ry = kotlin.math.round(y)
        if (rx == 50.0 && ry == 50.0) return null
        return "${rx.toInt()}% ${ry.toInt()}%"
    }

    fun defaultImagePrompt(title: String, description: String): String =
        """
        Generate an image for a course banner with the following title and description:
        Title: $title
        Description: $description
        """.trimIndent()

    fun defaultTimezone(): String = TimeZone.getDefault().id

    fun timezoneOptions(): List<String> = TimeZone.getAvailableIDs().sorted()

    fun contentPages(items: List<CourseStructureItem>): List<CourseStructureItem> =
        items.filter { it.kind == "content_page" }

    fun cacheKeySettings(courseCode: String): String = "course:$courseCode:settings"

    fun applyCourseToForm(course: CourseSummary): CourseGeneralFormState {
        val endParts = isoDurationToParts(course.relativeEndAfter)
        val hiddenParts = isoDurationToParts(course.relativeHiddenAfter)
        return CourseGeneralFormState(
            title = course.title,
            description = course.description,
            published = course.published == true,
            gradeLevel = course.gradeLevel.orEmpty(),
            courseTimezone = course.courseTimezone?.trim()?.takeIf { it.isNotEmpty() } ?: defaultTimezone(),
            courseHomeLanding = normalizeCourseHomeLanding(course.courseHomeLanding),
            courseHomeContentItemId = course.courseHomeContentItemId.orEmpty().trim(),
            scheduleMode = if (course.scheduleMode == ScheduleMode.relative.name) ScheduleMode.relative else ScheduleMode.fixed,
            startsAt = course.startsAt.orEmpty(),
            endsAt = course.endsAt.orEmpty(),
            visibleFrom = course.visibleFrom.orEmpty(),
            hiddenAt = course.hiddenAt.orEmpty(),
            relEndAmount = endParts.first,
            relEndUnit = endParts.second,
            relHiddenAmount = hiddenParts.first,
            relHiddenUnit = hiddenParts.second,
            markdownThemePreset = course.markdownThemePreset ?: "default",
            customDraft = course.markdownThemeCustom ?: MarkdownThemeCustom.seed(),
        )
    }

    fun buildCourseUpdateRequest(form: CourseGeneralFormState): CourseUpdateRequest {
        val mode = form.scheduleMode
        return CourseUpdateRequest(
            title = form.title.trim(),
            description = form.description.trim(),
            published = form.published,
            startsAt = if (mode == ScheduleMode.relative) null else form.startsAt.trim().ifEmpty { null },
            endsAt = if (mode == ScheduleMode.relative) null else form.endsAt.trim().ifEmpty { null },
            visibleFrom = if (mode == ScheduleMode.relative) null else form.visibleFrom.trim().ifEmpty { null },
            hiddenAt = if (mode == ScheduleMode.relative) null else form.hiddenAt.trim().ifEmpty { null },
            scheduleMode = mode.name,
            relativeEndAfter = if (mode == ScheduleMode.relative) partsToIsoDuration(form.relEndAmount, form.relEndUnit) else null,
            relativeHiddenAfter = if (mode == ScheduleMode.relative) partsToIsoDuration(form.relHiddenAmount, form.relHiddenUnit) else null,
            courseHomeLanding = form.courseHomeLanding.name,
            courseHomeContentItemId = if (form.courseHomeLanding == CourseHomeLanding.content_page) {
                form.courseHomeContentItemId.trim().ifEmpty { null }
            } else null,
            courseTimezone = form.courseTimezone.trim().ifEmpty { null },
            gradeLevel = form.gradeLevel.trim().ifEmpty { null },
        )
    }

    fun buildMarkdownThemePatch(form: CourseGeneralFormState): CourseMarkdownThemePatch =
        CourseMarkdownThemePatch(
            preset = form.markdownThemePreset,
            custom = if (form.markdownThemePreset == "custom") form.customDraft else null,
        )

    fun isGeneralFormDirty(form: CourseGeneralFormState, course: CourseSummary): Boolean =
        form != applyCourseToForm(course)

    fun courseNeedsUpdate(form: CourseGeneralFormState, course: CourseSummary): Boolean =
        buildCourseUpdateRequest(form) != buildCourseUpdateRequest(applyCourseToForm(course))

    fun themeNeedsUpdate(form: CourseGeneralFormState, course: CourseSummary): Boolean =
        buildMarkdownThemePatch(form) != buildMarkdownThemePatch(applyCourseToForm(course))

    fun isoToLocalDateString(iso: String?): String {
        if (iso.isNullOrBlank()) return ""
        val instant = runCatching { java.time.Instant.parse(iso) }.getOrNull() ?: return ""
        val zoned = instant.atZone(java.time.ZoneId.systemDefault())
        return java.time.format.DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm").format(zoned)
    }

    fun localDateStringToIso(value: String): String? {
        val trimmed = value.trim()
        if (trimmed.isEmpty()) return null
        return runCatching {
            val local = java.time.LocalDateTime.parse(trimmed, java.time.format.DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm"))
            local.atZone(java.time.ZoneId.systemDefault()).toInstant().toString()
        }.getOrNull()
    }
}

data class CourseGeneralFormState(
    val title: String = "",
    val description: String = "",
    val published: Boolean = false,
    val gradeLevel: String = "",
    val courseTimezone: String = CourseSettingsLogic.defaultTimezone(),
    val courseHomeLanding: CourseSettingsLogic.CourseHomeLanding = CourseSettingsLogic.CourseHomeLanding.data,
    val courseHomeContentItemId: String = "",
    val scheduleMode: CourseSettingsLogic.ScheduleMode = CourseSettingsLogic.ScheduleMode.fixed,
    val startsAt: String = "",
    val endsAt: String = "",
    val visibleFrom: String = "",
    val hiddenAt: String = "",
    val relEndAmount: String = "",
    val relEndUnit: CourseSettingsLogic.RelativeDurationUnit = CourseSettingsLogic.RelativeDurationUnit.M,
    val relHiddenAmount: String = "",
    val relHiddenUnit: CourseSettingsLogic.RelativeDurationUnit = CourseSettingsLogic.RelativeDurationUnit.M,
    val markdownThemePreset: String = "default",
    val customDraft: MarkdownThemeCustom = MarkdownThemeCustom.seed(),
)
